### DISCUSSION TITLE
Meshing error in defining my own cell model

### DISCUSSION THREAD

**Discussion by kumaryash7** [pybamm-team/PyBaMM#2809] [Q&A]
I have defined the negative electrode, separator, a positive electrode, and the positive electrode, negative and positive particles geometry , equations and parameters.

Geometry includes:-
geometry = {
    "separator": {x_s: {"min": L_n, "max": L_p}},        
    "positive electrode": {x_p: {"min": L_s, "max": L_p}},
    "negative electrode": {x_n: {"min": 0, "max": L_n}},
    "positive particle": {r_pos: {"min": 0, "max": R_p}},
    "negative particle": {r_neg: {"min": 0, "max": R_n}},
}

My code for the meshing part includes the following:-
submesh_types = {
    "separator": pybamm.Uniform1DSubMesh,
    "positive electrode": pybamm.Uniform1DSubMesh,
    "positive particle": pybamm.Uniform1DSubMesh,
    "negative electrode": pybamm.Uniform1DSubMesh,
    "negative particle": pybamm.Uniform1DSubMesh,
}

var_pts = {x_s: 10, x_p: 20,x_n:20, r_pos: 30,r_neg:30}
mesh = pybamm.Mesh(geometry, submesh_types, var_pts)

spatial_methods = {
    "separator": pybamm.FiniteVolume(),
    "positive electrode": pybamm.FiniteVolume(),
    "positive particle": pybamm.FiniteVolume(),
    "negative electrode": pybamm.FiniteVolume(),
    "negative particle": pybamm.FiniteVolume(),
}

disc = pybamm.Discretisation(mesh, spatial_methods)

disc.process_model(model);

After running the code till this point , I get the following error:-
submesh edges are not aligned

**Comment by valentinsulzer** [comment]
Geometry x positive min should be Ln + Ls

**Reply by kumaryash7** [reply]
geometry = {
    "separator": {x_s: {"min": L_n, "max": L_p}},        
    "positive electrode": {x_p: {"min": L_n+L_s, "max": L_p}},
    "negative electrode": {x_n: {"min": 0, "max": L_n}},
    "positive particle": {r_pos: {"min": 0, "max": R_p}},
    "negative particle": {r_neg: {"min": 0, "max": R_n}},
}

I have changed the geometry but still I am getting the same error that submeshes edges are not aligned

**Reply by valentinsulzer** [reply]
Change separator max to Ln + Ls. 

Learning to understand and debug python code will help you solve these problems faster

**Reply by kumaryash7** [reply]
I have changed the geometry and the meshing error is resolved
But now I am getting other error:-->

KeyError                                  Traceback (most recent call last)
File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:795, in Discretisation.process_symbol(self, symbol)
    794 try:
--> 795     return self._discretised_symbols[symbol]
    796 except KeyError:

KeyError: Subtraction(-0x129b72aaedf724f8, -, children=['150000.0 * 2.0 * 4.743416490252569e-10 * (boundary value(Positive particle concentration [mol.m-3]) ** 0.5) * ((51385.0 - boundary value(Positive particle concentration [mol.m-3])) ** 0.5) * sinh(19.461888122453146 * (Positive electrode potential [V] - Positive electrolyte potential [V] - (2.16216 + 0.07645 * tanh(30.834 - (0.0011259783438746717 * boundary value(Positive particle concentration [mol.m-3]))) + 2.1581 * tanh(52.294 - (0.001039451746618663 * boundary value(Positive particle concentration [mol.m-3]))) - (0.14169 * tanh(11.0923 - (0.0004103389432713827 * boundary value(Positive particle concentration [mol.m-3])))) + 0.2051 * tanh(1.4684 - (0.00011343982874379685 * boundary value(Positive particle concentration [mol.m-3]))) + 0.2531 * tanh(4.291641337386018 - (0.0001570479481286303 * boundary value(Positive particle concentration [mol.m-3]))) - (0.02167 * tanh(-87.5 + 0.0034445849956212904 * boundary value(Positive particle concentration [mol.m-3]))))))', 'div(3.8 * grad(Positive electrode potential [V]))'], domains={'primary': ['positive electrode']})

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:795, in Discretisation.process_symbol(self, symbol)
    794 try:
--> 795     return self._discretised_symbols[symbol]
    796 except KeyError:

KeyError: Divergence(0x32f9f5364039346c, div, children=['3.8 * grad(Positive electrode potential [V])'], domains={'primary': ['positive electrode']})

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:795, in Discretisation.process_symbol(self, symbol)
    794 try:
--> 795     return self._discretised_symbols[symbol]
    796 except KeyError:

KeyError: Multiplication(0x1bc6423b3f7600d4, *, children=['3.8', 'grad(Positive electrode potential [V])'], domains={'primary': ['positive electrode']})

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:795, in Discretisation.process_symbol(self, symbol)
    794 try:
--> 795     return self._discretised_symbols[symbol]
    796 except KeyError:

KeyError: Gradient(0x44b12638fe488712, grad, children=['Positive electrode potential [V]'], domains={'primary': ['positive electrode']})

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
Cell In[119], line 1
----> 1 disc.process_model(model);

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:225, in Discretisation.process_model(self, model, inplace, check_model, remove_independent_variables_from_rhs)
    223 # Process parabolic and elliptic equations
    224 pybamm.logger.verbose("Discretise model equations for {}".format(model.name))
--> 225 rhs, concat_rhs, alg, concat_alg = self.process_rhs_and_algebraic(model)
    226 model_disc.rhs, model_disc.concatenated_rhs = rhs, concat_rhs
    227 model_disc.algebraic, model_disc.concatenated_algebraic = alg, concat_alg

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:637, in Discretisation.process_rhs_and_algebraic(self, model)
    634 processed_concatenated_rhs = self._concatenate_in_order(processed_rhs)
    636 # Discretise and concatenate algebraic equations
--> 637 processed_algebraic = self.process_dict(model.algebraic)
    639 # Concatenate algebraic into a single state vector
    640 # Need to concatenate in order as the ordering of equations could be different
    641 # in processed_algebraic and model.algebraic
    642 processed_concatenated_algebraic = self._concatenate_in_order(
    643     processed_algebraic
    644 )

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:764, in Discretisation.process_dict(self, var_eqn_dict, ics)
    761         eqn = pybamm.FullBroadcast(eqn, broadcast_domains=eqn_key.domains)
    763 pybamm.logger.debug("Discretise {!r}".format(eqn_key))
--> 764 processed_eqn = self.process_symbol(eqn)
    766 # Calculate scale if the key has a scale
    767 scale = getattr(eqn_key, "scale", 1)

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:797, in Discretisation.process_symbol(self, symbol)
    795     return self._discretised_symbols[symbol]
    796 except KeyError:
--> 797     discretised_symbol = self._process_symbol(symbol)
    798     self._discretised_symbols[symbol] = discretised_symbol
    799     discretised_symbol.test_shape()

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:833, in Discretisation._process_symbol(self, symbol)
    831 left, right = symbol.children
    832 disc_left = self.process_symbol(left)
--> 833 disc_right = self.process_symbol(right)
    834 if symbol.domain == []:
    835     return pybamm.simplify_if_constant(
    836         symbol._binary_new_copy(disc_left, disc_right)
    837     )

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:797, in Discretisation.process_symbol(self, symbol)
    795     return self._discretised_symbols[symbol]
    796 except KeyError:
--> 797     discretised_symbol = self._process_symbol(symbol)
    798     self._discretised_symbols[symbol] = discretised_symbol
    799     discretised_symbol.test_shape()

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:861, in Discretisation._process_symbol(self, symbol)
    858 elif isinstance(symbol, pybamm.UnaryOperator):
    859     child = symbol.child
--> 861     disc_child = self.process_symbol(child)
    862     if child.domain != []:
    863         child_spatial_method = self.spatial_methods[child.domain[0]]

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:797, in Discretisation.process_symbol(self, symbol)
    795     return self._discretised_symbols[symbol]
    796 except KeyError:
--> 797     discretised_symbol = self._process_symbol(symbol)
    798     self._discretised_symbols[symbol] = discretised_symbol
    799     discretised_symbol.test_shape()

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:833, in Discretisation._process_symbol(self, symbol)
    831 left, right = symbol.children
    832 disc_left = self.process_symbol(left)
--> 833 disc_right = self.process_symbol(right)
    834 if symbol.domain == []:
    835     return pybamm.simplify_if_constant(
    836         symbol._binary_new_copy(disc_left, disc_right)
    837     )

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:797, in Discretisation.process_symbol(self, symbol)
    795     return self._discretised_symbols[symbol]
    796 except KeyError:
--> 797     discretised_symbol = self._process_symbol(symbol)
    798     self._discretised_symbols[symbol] = discretised_symbol
    799     discretised_symbol.test_shape()

File ~\anaconda3\lib\site-packages\pybamm\discretisations\discretisation.py:866, in Discretisation._process_symbol(self, symbol)
    863     child_spatial_method = self.spatial_methods[child.domain[0]]
    865 if isinstance(symbol, pybamm.Gradient):
--> 866     return child_spatial_method.gradient(child, disc_child, self.bcs)
    868 elif isinstance(symbol, pybamm.Divergence):
    869     return child_spatial_method.divergence(child, disc_child, self.bcs)

File ~\anaconda3\lib\site-packages\pybamm\spatial_methods\finite_volume.py:120, in FiniteVolume.gradient(self, symbol, discretised_symbol, boundary_conditions)
    118     bcs = boundary_conditions[symbol]
    119     if any(bc[1] == "Neumann" for bc in bcs.values()):
--> 120         out = self.add_neumann_values(symbol, out, bcs, domain)
    122 return out

File ~\anaconda3\lib\site-packages\pybamm\spatial_methods\finite_volume.py:766, in FiniteVolume.add_neumann_values(self, symbol, discretised_gradient, bcs, domain)
    763 n = submesh.npts - 1
    764 second_dim_repeats = self._get_auxiliary_domain_repeats(symbol.domains)
--> 766 lbc_value, lbc_type = bcs["left"]
    767 rbc_value, rbc_type = bcs["right"]
    769 # Count number of Neumann boundary conditions

KeyError: 'left'
